### DISCUSSION TITLE
Flipping Crazyflies

### DISCUSSION THREAD

**Discussion by uhvardhan** [USC-ACTLab/crazyswarm#672] [Q&A]
Hello,

I am trying to use vicon motion capture + crazyswarm. 

I am using the following configurations: Duplicate marker arrangement and "libobjecttracker" was being used.

```
 <?xml version="1.0"?>
<launch>
  <arg name="joy_dev" default="/dev/input/js0" />

  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflieTypes.yaml" />
  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflies.yaml" /> 

  <node pkg="crazyswarm" type="crazyswarm_server" name="crazyswarm_server" output="screen" >
    <rosparam>
      # Logging configuration (Use enable_logging to actually enable logging)
      genericLogTopics: ["log1"]
      genericLogTopicFrequencies: [10]
      genericLogTopic_log1_Variables: ["stateEstimate.x", "ctrltarget.x"]
      # firmware parameters for all drones (use crazyflieTypes.yaml to set per type, or
      # allCrazyflies.yaml to set per drone)
      firmwareParams:
        commander:
          enHighLevel: 1
        stabilizer:
          estimator: 2 # 1: complementary, 2: kalman
          controller: 2 # 1: PID, 2: mellinger
        ring:
          effect: 16 # 6: double spinner, 7: solid color, 16: packetRate
          solidBlue: 255 # if set to solid color
          solidGreen: 0 # if set to solid color
          solidRed: 0 # if set to solid color
          headlightEnable: 0
        locSrv:
          extPosStdDev: 1e-3
          extQuatStdDev: 0.5e-1
        kalman:
          resetEstimation: 1
      # tracking
      motion_capture_type: "vicon"
      object_tracking_type: "libobjecttracker" # one of motionCapture,libobjecttracker
      send_position_only: False # set to False to send position+orientation; set to True to send position only
      motion_capture_host_name: "192.168.1.1"
      # motion_capture_interface_ip: "" # optional for optitrack with multiple interfaces
      save_point_clouds: "/dev/null" # set to a valid path to log mocap point cloud binary file.
      print_latency: True
      write_csvs: False
      force_no_cache: False
      enable_parameters: True
      enable_logging: True
      enable_logging_pose: False
    </rosparam>
  </node>

  <node name="joy" pkg="joy" type="joy_node" output="screen">
    <param name="dev" value="$(arg joy_dev)" />
  </node>

  <node pkg="crazyswarm" type="crazyswarm_teleop" name="crazyswarm_teleop" output="screen">
    <param name="csv_file" value="$(find crazyswarm)/launch/figure8_smooth.csv" />
    <param name="timescale" value="0.8" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find crazyswarm)/launch/test.rviz"/>

  <!-- <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot_x" args="/cf2/log1/values[0]"/> -->
  <!-- <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot_roll" args="/cf1/log1/values[2] /cf1/log1/values[3]"/> -->

</launch>
```

I am able to run 3 cf's on channel 86 using only 1 crazyradio PA. However when I add the 4th one, atleast one of them flips. The flipping happens at the start of running figure8.py . I have attached a video for reference : https://youtu.be/9_BlTbOxvBc

I chose channel 86 by running radio-test.py file. Even if I add the 4th cf on different radio, the flipping still happens. Not really sure where the issue is.

Thanks - @whoenig @jpreiss - Sorry. Been stuck on this for about 4 days now.

**Comment by whoenig** [comment]
Unless you really need logging, I suggest that you disable it. Your radio bandwidth might saturate which can cause the crazyswarm to send stale pose information to the drones.

When you "add the 4th cf on different radio", I assume that you mean that you select a different channel for the 4th one? Note that the channels need to be spaced apart by at least 2, if you use the 2M mode.

For Crazyflie 2.1 and "new setups" I recommend the single-marker tracking. It's a bit more robust and uses less radio bandwidth.

**Reply by uhvardhan** [reply]
Hi @whoenig,

Thankyou very much for the reply. 

I am going to try out the single-marker tracking today. Hopefully I will be able to fly 20 of them. 

Yes. “add the 4th cf on different radio” means that I am choosing a different channel (I chose 72) and a different radio.

**Reply by uhvardhan** [reply]
I would also like to mention one another issue. 

When I try to flash the cf using chooser.py - I have the following issues:

`WARNING: Could not find STM32 firmware (/home/*******/github_files/crazyswarm/ros_ws/src/crazyswarm/scripts/../../../../crazyflie-firmware/cf2.bin).
WARNING: Could not find NRF51 firmware (/home/*****/github_files/crazyswarm/ros_ws/src/crazyswarm/scripts/../../../../crazyflie2-nrf-firmware/cf2_nrf.bin).
`

Any fix for this?

**Reply by whoenig** [reply]
You would need to compile the firmware you want to use yourself (in the crazyflie-firmware and crazyflie2-nrf-firmware folders).

**Comment by uhvardhan** [comment]
Hi @whoenig,

So I have tried the approach you have mentioned. I am also attaching a video for your reference: https://youtu.be/Uwbj3WpK-uc

I used a single marker only for each cf. The first four cf are using the channel 82. The 5th cf which failed uses the channel 86. I have ensured that crazyradio PA is version 99.55. One crazyradio is connected via usb on the laptop. Another one is connected via Anker USB deck.  

I have updated the firmware of cf using cfclient. I am not sure if it is going to affect the system. I have also made the following changes:
`
send_position_only: True
enable_logging: False
enable_logging_pose: False 
`
There have been instances where cf5 flew but cf4 failed. Sometimes, it’s random.

Thanks.

**Reply by whoenig** [reply]
Thanks for the feedback. Your setup you describe should be fine (both in terms on how the radios are attached and settings in the launch file).

Does the problem also occur if you use a single radio? With logging disabled, you should be able to control 20 drones or so with single-marker tracking...

Also, I assume you use a native installation (not a VM) and the motion capture and control PC are communicating in a wired network?

**Reply by uhvardhan** [reply]
Thank you for the reply. 

Yes. The issue exists even with a single radio. I am using Ubuntu 20.04 on my laptop. The Ubuntu laptop and motion capture pc(where tracker is only ON) are communicating via wired network.

**Reply by uhvardhan** [reply]
So, I have an update. On a different computer, I did a fresh Ubuntu 20.04 install + ROS + crazyswarm. I was able to fly about 10 with single marker arrangement on 1 crazy radio on channel 82. However I am unable to do so consistently i.e., multiple times. I have disabled all logging. 

Any suggestions?

**Reply by whoenig** [reply]
- When you say it is inconsistent, it means that in some cases you get the same crashes as before?
- The best approach for debugging would be to use the uSD-card deck and log the data that you receive on-board. Do you have that add-on board + uSD cards? The most common issue is that the motion capture has spurious markers in some of the frames.

**Reply by uhvardhan** [reply]
Hi @whoenig, 

Sorry for the delay in responding. 

Yes. When I say inconsistent, I mean at least 1 cf crashes every trial. I unfortunately do not have the uSD-card deck. I ensured there are no spurious markers. 

Update: I have been able to reliably (meaning fly all the cf for 4-5 tests one after another) fly 9 cf's [single marker arrangement] only. If I can reach 15 I would be happy. 

Thanks.

**Reply by uhvardhan** [reply]
A general observation: PID works more reliably than Mellinger. All the work I have described above is with PID.

**Reply by uhvardhan** [reply]
Regarding the crazyradio, if I am flying cfs at a maximum height of 2.5m, is there any specific height I should keep the crazyradio at?

**Reply by whoenig** [reply]
So the only change in order to get the 9 working was to change the controller to PID?

Our default gains for Mellinger are pretty high. If there are stability issues, this usually means that either the motion capture tracking is very noisy, or that your drones fly pretty close to each other. In the first case, you can try reducing the EKF stddev, use bigger markers, and make sure you re-calibrate your mocap frequently. In the second case, you would need to check if spacing the drones more apart helps. We found that one needs about 0.3m horizontally and 0.6 vertically to avoid too strong interactions.

The radio placement is usually not a problem for indoor operation. If you suspect radio issues (check with the com_test application of crazyflie_tools), it is mostly caused by the channel being used by other 2.4 Ghz devices. Flying the CFs high can sometimes cause tracking issues, depending on the placement of the cameras. For flying 2.5m, you probably need the cameras to be mounted at around 3.5m height.

**Reply by uhvardhan** [reply]
Hi @whoenig - Yes. The only change I did to get 9 working was changing the controller to PID. 
Where can I reduce the EKF stddev? I am currently using a single 14mm marker on each cf. Should I go bigger? I will re-calibrate the mocap frequently. I usually space all cfs at least 0.5m apart. So not really sure if I should space the drones more. 

So, I could not find com_test application. I can see comCheck but wasn't able to run it. 

I have cameras at about 2.3m-2.4m high. I am trying to just get the cf's to hover first.
